File: app/src/test/java/at/tomtasche/reader/background/StreamUtilTest.java
package at.tomtasche.reader.background;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import org.junit.Assert;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;

public class StreamUtilTest {

    @Rule public TemporaryFolder temporaryFolder = new TemporaryFolder();

    @Test
    public void copyStreamToStream() throws IOException {
        byte[] data = "hello world".getBytes(StandardCharsets.UTF_8);
        ByteArrayOutputStream out = new ByteArrayOutputStream();

        StreamUtil.copy(new ByteArrayInputStream(data), out);

        Assert.assertArrayEquals(data, out.toByteArray());
    }

    @Test
    public void copyLargeStream() throws IOException {
        // larger than the internal 1024 byte buffer
        byte[] data = new byte[10000];
        for (int i = 0; i < data.length; i++) {
            data[i] = (byte) (i % 256);
        }
        ByteArrayOutputStream out = new ByteArrayOutputStream();

        StreamUtil.copy(new ByteArrayInputStream(data), out);

        Assert.assertArrayEquals(data, out.toByteArray());
    }

    @Test
    public void copyEmptyStream() throws IOException {
        ByteArrayOutputStream out = new ByteArrayOutputStream();

        StreamUtil.copy(new ByteArrayInputStream(new byte[0]), out);

        Assert.assertEquals(0, out.size());
    }

    @Test
    public void copyStreamToFile() throws IOException {
        byte[] data = "file content".getBytes(StandardCharsets.UTF_8);
        File target = temporaryFolder.newFile();

        StreamUtil.copy(new ByteArrayInputStream(data), target);

        Assert.assertArrayEquals(data, Files.readAllBytes(target.toPath()));
    }

    @Test
    public void copyFileToFile() throws IOException {
        byte[] data = "copy me".getBytes(StandardCharsets.UTF_8);
        File source = temporaryFolder.newFile();
        Files.write(source.toPath(), data);
        File target = temporaryFolder.newFile();

        StreamUtil.copy(source, target);

        Assert.assertArrayEquals(data, Files.readAllBytes(target.toPath()));
    }

    @Test
    public void readFullyReturnsUtf8String() throws IOException {
        String text = "umlauts: äöü, emoji: 😀";
        InputStream in = new ByteArrayInputStream(text.getBytes(StandardCharsets.UTF_8));

        Assert.assertEquals(text, StreamUtil.readFully(in));
    }

    @Test
    public void copyClosesInputStream() throws IOException {
        File source = temporaryFolder.newFile();
        Files.write(source.toPath(), "x".getBytes(StandardCharsets.UTF_8));
        FileInputStream in = new FileInputStream(source);

        StreamUtil.copy(in, new ByteArrayOutputStream());

        try {
            in.read();
            Assert.fail("input stream should be closed after copy");
        } catch (IOException expected) {
            // expected
        }
    }
}

File: app/src/test/java/app/opendocument/droid/background/StreamUtilTest.kt
package app.opendocument.droid.background

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStream
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import org.junit.Assert
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TemporaryFolder

class StreamUtilTest {

    @get:Rule val temporaryFolder = TemporaryFolder()

    @Test
    fun copyStreamToStream() {
        val data = "hello world".toByteArray(StandardCharsets.UTF_8)
        val out = ByteArrayOutputStream()

        StreamUtil.copy(ByteArrayInputStream(data), out)

        Assert.assertArrayEquals(data, out.toByteArray())
    }

    @Test
    fun copyLargeStream() {
        // larger than the internal 1024 byte buffer
        val data = ByteArray(10000)
        for (i in data.indices) {
            data[i] = (i % 256).toByte()
        }
        val out = ByteArrayOutputStream()

        StreamUtil.copy(ByteArrayInputStream(data), out)

        Assert.assertArrayEquals(data, out.toByteArray())
    }

    @Test
    fun copyEmptyStream() {
        val out = ByteArrayOutputStream()

        StreamUtil.copy(ByteArrayInputStream(ByteArray(0)), out)

        Assert.assertEquals(0, out.size())
    }

    @Test
    fun copyStreamToFile() {
        val data = "file content".toByteArray(StandardCharsets.UTF_8)
        val target = temporaryFolder.newFile()

        StreamUtil.copy(ByteArrayInputStream(data), target)

        Assert.assertArrayEquals(data, Files.readAllBytes(target.toPath()))
    }

    @Test
    fun copyFileToFile() {
        val data = "copy me".toByteArray(StandardCharsets.UTF_8)
        val source = temporaryFolder.newFile()
        Files.write(source.toPath(), data)
        val target = temporaryFolder.newFile()

        StreamUtil.copy(source, target)

        Assert.assertArrayEquals(data, Files.readAllBytes(target.toPath()))
    }

    @Test
    fun readFullyReturnsUtf8String() {
        val text = "umlauts: äöü, emoji: 😀"
        val input: InputStream = ByteArrayInputStream(text.toByteArray(StandardCharsets.UTF_8))

        Assert.assertEquals(text, StreamUtil.readFully(input))
    }

    @Test
    fun copyClosesInputStream() {
        val source = temporaryFolder.newFile()
        Files.write(source.toPath(), "x".toByteArray(StandardCharsets.UTF_8))
        val input = FileInputStream(source)

        StreamUtil.copy(input, ByteArrayOutputStream())

        try {
            input.read()
            Assert.fail("input stream should be closed after copy")
        } catch (expected: IOException) {
            // expected
        }
    }
}
